### ISSUE TITLE
Curious issue in Eclipse debugging

### ISSUE THREAD

**Issue by marcieldeg** [FasterXML/jackson-databind#3198]
Please check this code:

```
import com.fasterxml.jackson.databind.ObjectMapper;

public class Main {
	public static void main(String[] args) throws Exception {
		Boolean bool = Boolean.FALSE;
		if (bool.booleanValue())
			new ObjectMapper().readValue("{\"stringVal\":\"aaa\",\"longVal\":123,\"boolVal\":true}", Test.class);
	}

	public static class Test {
		public String stringVal;
		public Long longVal;
		public Boolean boolVal;

		public Test() {
		}
	}
}
```

1. Put a breakpoint on line `if (bool.booleanValue())` and run in debug mode on eclipse.
2. Using the tab "Variables" from Eclipse, change internal value from variable "bool" to "true":
![image](https://user-images.githubusercontent.com/14255958/124477672-a8a3fc80-dd7a-11eb-9351-02440a0a80f1.png)
3. Click "Resume" (F8). It will cause a Jackson error: 
`Exception in thread "main" com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `testes.Main$Test` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)`.
It's not right as there is a constructor in the "Test" class.

4. Repeat the test, changing the variable "bool" value using the option "Change Value..." (Right click), as in the image below:
![image](https://user-images.githubusercontent.com/14255958/124478417-8b236280-dd7b-11eb-9d11-9887249cc92e.png)
5. In this case, it works well.

**Comment by cowtowncoder** [comment]
Sounds like an Eclipse oddity and something I have not seen before.
Leaving open in case anyone else knows -- but I would suggest trying mailing list

https://groups.google.com/g/jackson-user

first since it will probably reach more developers who might have seen something like this before.

**Comment by Dragas** [comment]
You've changed how JVM works. When you rewrote boolean value of constant `Boolean.FALSE`, any places that depend on wrapped boolean value will now incorrectly treat `FALSE` values as if they were primitive true. This is the same issue as breaking maths when rewriting values in `Integer.IntegerCache.cache` field. See https://stackoverflow.com/questions/63733340/how-to-modify-integer-addition-in-java

**Comment by cowtowncoder** [comment]
Whoa. Did not know Eclipse tooling can achieve that... (I was aware of canonical instances for `Integer`, `Long` and `Boolean`).

**Comment by Dragas** [comment]
It's general usecase of any debugger. You can rewrite values at runtime as you please. You can achieve the same with reflection
see example in https://stackoverflow.com/a/3301720/6523288

**Comment by cowtowncoder** [comment]
Yes, I can see how Reflection approach would work; it is often necessary to at least access internal fields (less often to change). I am just bit surprised that Eclipse debugger does not have some guard rails in a few cases like this.
Although arguably a set of dangerous cases is bit unlimited so there's a trade-off.

Thank you for solving this mystery @Dragas !
